test/reduce.js

```diff
@@ -116,6 +116,11 @@ module.exports = function reduce_test(testcase, minify_options, reduce_options)
                 ][ ((node.start._permute += step) * steps | 0) % 3 ];
             }
             else if (node instanceof U.AST_BlockStatement) {
+                if (in_list) {
+                    node.start._permute++;
```

Comment:
leaving _permute as in integer should be fine since there are only 2 permutation cases for AST_BlockStatement. There's a chance that subsequent code later on might bump the _permute count though, in which case you'd have to change this to use fractional steps. Try it out and see if it reduces the original test case for https://github.com/mishoo/UglifyJS2/pull/3705#issuecomment-582927572 further.

Reply:
2 is `REPLACEMENTS.length`, by the way. The max limit of _permute.

Reply:
The specified test case now reduced to:
```
$ uglifyjs test.js -c --reduce-test
// reduce test pass 1, iteration 0: 7743 bytes
// reduce test pass 1, iteration 25: 3869 bytes
// reduce test pass 1, iteration 50: 3863 bytes
// reduce test pass 1, iteration 75: 3626 bytes
// reduce test pass 1, iteration 100: 2635 bytes
// reduce test pass 1, iteration 125: 2583 bytes
// reduce test pass 1, iteration 150: 2558 bytes
// reduce test pass 1, iteration 175: 1779 bytes
// reduce test pass 1, iteration 200: 1208 bytes
// reduce test pass 1: 245 bytes
// reduce test pass 2: 206 bytes
// reduce test pass 3: 206 bytes
var c = 0;

var expr1 = this;

for (var key1 in expr1) {
    var bar_2 = expr1[key1];
    var expr2 = bar_2;
    for (var key2 in expr2) {
        c = c + 1;
    }
}

var b = 0;

switch (0) {
  case 0:
    var a = 0;
    var bar_2 = 0;
}

console.log(null, a, b, c, Infinity, NaN, undefined);
// output: null 0 0 21 Infinity NaN undefined
// minify: null 0 0 17 Infinity NaN undefined
// options: {"compress":true,"mangle":false}
```

Reply:
👍

Reply:
even better... with this latest PR:
```
$ bin/uglifyjs 6355.js -c toplevel --reduce-test
// reduce test pass 1, iteration 0: 7680 bytes
// reduce test pass 1, iteration 25: 3869 bytes
// reduce test pass 1, iteration 50: 3851 bytes
// reduce test pass 1, iteration 75: 3652 bytes
// reduce test pass 1, iteration 100: 3626 bytes
// reduce test pass 1, iteration 125: 2635 bytes
// reduce test pass 1, iteration 150: 2568 bytes
// reduce test pass 1, iteration 175: 2548 bytes
// reduce test pass 1, iteration 200: 1615 bytes
// reduce test pass 1: 233 bytes
// reduce test pass 2: 138 bytes
// reduce test pass 3: 115 bytes
var c = 0;

var expr1 = this;

for (var key1 in expr1) {
    c = c + 1;
}

var b = 0;

var a = 0;

console.log(null, a, b, c, Infinity, NaN, undefined);
// output: null 0 0 5 Infinity NaN undefined
// minify: null 0 0 2 Infinity NaN undefined
// options: {"compress":{"toplevel":true},"mangle":false}
```
which is as good as you can hope to get.